### DISCUSSION TITLE
Extending getModelForClass

### DISCUSSION THREAD

**Discussion by ksromero** [typegoose/typegoose#688] [Q&A]
## Versions

- System: windows 
- NodeJS: 14.17.2
- Typescript: 4.6.3
- Compiler / Transpiler: ts-node-dev 
- Typegoose(NPM): 9.7.1 (@typegoose/typegoose) 
- mongoose: 6.2.10 
- mongodb: 5.0.7

---

Supposed I have a model.

```ts
import { getModelForClass, prop, Ref } from '@typegoose/typegoose'

class User {
  ..........
}

export default getModelForClass(User)
```

How can I extend the User model
```ts

import UserModel from '../models/User'

class UserRepository extends UserModel {

}

export default UserRepository
```
It says '**Base constructors must all have the same return type**.' when extending it, Is there a way where I can extend it ? Thanks

**Comment by hasezoey** [comment]
> How can I extend the User model

if you want to extend it, not just use it, you will need to extend the original class and then generate another model with another call to `getModelForClass`

like:
```ts
class Animal {
  @prop()
  public name?: string;
}

const AnimalModel = getModelForClass(Animal);

class Cat extends Animal {
  @prop()
  public purrsOften?: boolean;
}

const CatModel = getModelForClass(Cat);
```

**Reply by ksromero** [reply]
No, what i mean is extend it like from a normal class (doesnt use a getModelForClass), not another model 

```
class NormalClass extends User {

}
```

**Reply by hasezoey** [reply]
> No, what i mean is extend it like from a normal class (doesnt use a getModelForClass), not another model

sorry, but i have no clue what you are asking about

**Reply by smolinari** [reply]
@hasezoey - I think he's looking for something like instance methods, maybe? 

https://typegoose.github.io/typegoose/docs/guides/quick-start-guide/#instance-methods

@ksromero - You can do one of two things to support something like a repository pattern (which it seems you want). You can inject the model into the repo class, and build out your own repository using the default model methods. Mongoose's model is quite powerful. Or,  you can add instance methods to the schema (and ultimately to the model) and just continue to use the model (no repository necessary). 

Scott

**Reply by ksromero** [reply]
Yeah , my flow is I will extend the class of getModelForClass in the repository, in repository is where I put some of the custom query I have and inject the repo on the service but i cant seem to extend the getModelForClass in my repository class

**Reply by smolinari** [reply]
@ksromero - You can't extend `getModelForClass`, at least not for the purposes you are looking for. You either add instance methods to your model's schema or you create a service-like class that injects the model and acts as your repository. 

Scott
